### PR TITLE
Add Armenia [work in progress]

### LINKED ISSUE #1253: Data for Armenia AM
There is a generation **schedule** for Armenia with data for hydro, gas and nuclear:

It is in the PDF in the second link on this side (ends with aktiv.pdf)

http://www.energyoperator.am/information.html

I ran the PDF through an OCR converter for fun do decipher the first column. Worked perfectly for letters and numbers, just the "-" signs had to be replaced.

I attached an [Excel file](https://github.com/tmrowco/electricitymap/files/1833243/armenia.xlsx) with the outcomes and some edits to highlight the rows of interest.

Here is what the result looks like:

image

Even if this is only a day-ahead schedule, it could be worth downloading the PDF each day just to have them, since there seems to be no data for the previous days or any historical data.

### PR DESCRIPTION
**Pull request by alixunderplatz**
The basic parser for Armenia 🇦🇲 is done :) High data frequency:  updated every minute 👍 👍 👍 
Thanks for the support @corradio! 👍
 
Now I need you guys to **carefully check and to improve** it! 😄 
Feel free to add your tweaks directly or comment the lines with your suggestions.
**Please add a list of what still has to be adjusted in other files to get this thing working ;)**

_______________________________
Sample output:

```
fetch_production ->
{'zoneKey': 'AM', 'datetime': datetime.datetime(2019, 4, 18, 23, 59, tzinfo=tzfile('Asia/Yerevan')), 'production': {'gas': 225.0, 'hydro': 451.0, 'nuclear': 224.0, 'coal': 0, 'geothermal': 0, 'oil': 0}, 'storage': {'hydro storage': 0, 'battery storage': 0}, 'source': 'http://epso.am/poweren.htm'}
fetch_exchange(AM, AZ) ->
{'datetime': datetime.datetime(2019, 4, 18, 23, 59, tzinfo=tzfile('Asia/Yerevan')), 'netFlow': -17.5, 'source': 'http://epso.am/poweren.htm', 'sortedZoneKeys': 'AM->AZ'}
fetch_exchange(AM, GE) ->
{'datetime': datetime.datetime(2019, 4, 18, 23, 59, tzinfo=tzfile('Asia/Yerevan')), 'netFlow': -0.0, 'source': 'http://epso.am/poweren.htm', 'sortedZoneKeys': 'AM-GE'}
fetch_exchange(AM, IR) ->
{'datetime': datetime.datetime(2019, 4, 18, 23, 59, tzinfo=tzfile('Asia/Yerevan')), 'netFlow': 266.5, 'source': 'http://epso.am/poweren.htm', 'sortedZoneKeys': 'AM->IR'}
```
_______________________________
There is one problem with the timestamp I just got aware of:
The html only delivers the time (HH:mm) of the data, but not the date. This time is passed to time_string using dateutil, which automatically sets the timestamp's date.
"00:25" -> 2019-04-18 00:25:00
The point is, my machine's local time is used instead of Armenia's time. That's why it returns the 18th of April, although its the 19th in Armenia already.
Someone needs to fix the merge of time_string with the correct current date in Armenia. Or we simply use arrow.now(tz='Asia/Yerevan'), the difference is like 1 minute...
_______________
I make the request from the same URL, with the same lines of code two seperate times, in fetch_production and fetch_exchange, although one requst should be sufficient ;)
It's probably better to make the request for all the data in one step, because sometimes the connection closes, causing an error (maybe some error handling at this point: retry to reach the data url after 3 seconds or so...). This could lead to data gaps for either exchange or production, I guess?
_______________

Basically I extract the desired info and then the numbers from the html.
Here is what the souped html part of interest looks like (see data_string in my code and see mappings for variable description):

```
var Lalvar = "0"
var ninoc = "0"
var alaver = "0"
var shin = "-4.2"
var arcakh = "21.7"
var ahar = "-266.5"
var cons = "901"
var altern = "0"
var atom = "224"
var tes = "225"
var ges = "451"
var peretok = "-249.00"
var herc2 = "50.01"
var time2 = "23:59"
var sparum2 = "650"
```
__________________________
last but not least: Thanks to all you parser-writers for your code snippets ;) & to all q&a on stackoverflow and similar pages!
_________________________
Closes #1253 
"If it looks stupid, but it works, it ain't stupid" 😆

### CODE REVIEW

**Comment by tmslaine** [comment]
Aren't alternative stations hydro?
On http://epso.am/scada.htm they are listed as the last two entries under
Հիդրոկայ աններ / Hydro, on the right box in the middle of the first page,
with the same affix(I think GES or something similar) as the left column in that box

**Comment by alixunderplatz** [comment]
@tmslaine you're totally right! I forgot to check with hydro from the scada table.
I'll fix it in the parser, thank you! :)

**Comment by systemcatch** [comment]
Works for me on the commandline.
```shell
(EM-env) chris@ThinkPad:~/electricitymap$ python parsers/AM.py 
fetch_production->
{'zoneKey': 'AM', 'production': {'gas': 223.0, 'geothermal': 0, 'coal': 0, 'oil': 0, 'hydro': 332.0, 'nuclear': 228.0}, 'source': 'http://epso.am/poweren.htm', 'datetime': datetime.datetime(2019, 4, 26, 13, 14, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Yerevan')), 'storage': {'hydro storage': 0, 'battery storage': 0}}
fetch_exchange(AM, AZ) ->
{'sortedZoneKeys': 'AM->AZ', 'source': 'http://epso.am/poweren.htm', 'datetime': datetime.datetime(2019, 4, 26, 13, 14, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Yerevan')), 'netFlow': -20.3}
fetch_exchange(AM, GE) ->
{'sortedZoneKeys': 'AM-GE', 'source': 'http://epso.am/poweren.htm', 'datetime': datetime.datetime(2019, 4, 26, 13, 14, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Yerevan')), 'netFlow': 1.5}
fetch_exchange(AM, IR) ->
{'sortedZoneKeys': 'AM->IR', 'source': 'http://epso.am/poweren.htm', 'datetime': datetime.datetime(2019, 4, 26, 13, 14, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Yerevan')), 'netFlow': 4.6}
```
For the storage they can just be called 'hydro' & 'battery', we're sure it's zero?

File: parsers/AM.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    gas_total = float(gas_tes[0])
+
+    hydro_ges = re.findall("[-+]?[.]?[\d]+(?:,\d\d\d)*[\.]?\d*(?:[eE][-+]?\d+)?", data_split[11])
+    hydro_altern = re.findall("[-+]?[.]?[\d]+(?:,\d\d\d)*[\.]?\d*(?:[eE][-+]?\d+)?", data_split[8])
+    hydro_total = float(hydro_ges[0])+float(hydro_altern[0])
+
+    nuclear_atom = re.findall("[-+]?[.]?[\d]+(?:,\d\d\d)*[\.]?\d*(?:[eE][-+]?\d+)?", data_split[9])
+    nuclear_total = float(nuclear_atom[0])
+
+    time_string = dparser.parse(data_split[14].split()[3],fuzzy=True)
```
Comment (systemcatch) [inline parsers/AM.py]: We could do something like this to ensure the local time in Armenia is used, need to add a datetime import as well.
```suggestion
    yerevan = tz.gettz('Asia/Yerevan')
    date_time = dparser.parse(data_split[14].split()[3], default=datetime.now(yerevan), fuzzy=True)
```

File: parsers/AM.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    date_time = time_string.replace(tzinfo=tz.gettz('Asia/Yerevan'))
+
+    #Operating solar, wind and biomass plants exist in small numbers, but are not reported yet
```
Comment (systemcatch) [inline parsers/AM.py]: They can be added as None keys if they're present but unknown.

File: parsers/AM.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    IR_1 = re.findall("[-+]?[.]?[\d]+(?:,\d\d\d)*[\.]?\d*(?:[eE][-+]?\d+)?", data_split[6])
+
+    AM_AZ = float(AZ_1[0])+float(AZ_2[0])
+    AM_GE = float(GE_1[0])+float(GE_2[0])+float(GE_3[0])
+    AM_IR = float(IR_1[0])
+
+    time_string = dparser.parse(data_split[14].split()[3],fuzzy=True)
+    date_time = time_string.replace(tzinfo=tz.gettz('Asia/Yerevan'))
+ 
+    if '->'.join(sorted([zone_key1, zone_key2])) == 'AM->AZ':
```
Comment (systemcatch) [inline parsers/AM.py]: you can do;

```python
sorted_keys = '->'.join(sorted([zone_key1, zone_key2]))
```
at the start of the function which lets you avoid all these if statements.

File: parsers/AM.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+# coding=utf-8
+
+import arrow
+import re
+import requests
+from dateutil import parser as dparser
```
Comment (systemcatch) [inline parsers/AM.py]:
```suggestion
from dateutil import parser as dparser, tz
```

**Comment by systemcatch** [comment]
Things to do:

- [ ] Handle connection drops.
- [ ] Update exchanges config.
- [ ] Update readme.
- [ ] Test on mockserver.

**Comment by alixunderplatz** [comment]
@systemcatch Thank you for your suggestions! :) Can you check again, please?
____________________________________________
Sample output:
```
fetch_production->
{'zoneKey': 'AM', 'datetime': datetime.datetime(2019, 4, 28, 14, 35, 20, 212570, tzinfo=tzfile('Asia/Yerevan')), 'production': {'gas': 215.0, 'hydro': 314.0, 'nuclear': 227.0, 'biomass': None, 'coal': 0, 'geothermal': 0, 'oil': 0, 'solar': None, 'wind': None}, 'storage': {'hydro': 0, 'battery': 0}, 'source': 'http://epso.am/poweren.htm'}
fetch_exchange(AM, AZ) ->
{'sortedZoneKeys': 'AM->AZ', 'datetime': datetime.datetime(2019, 4, 28, 14, 35, 23, 931060, tzinfo=tzfile('Asia/Yerevan')), 'netFlow': -28.7, 'source': 'http://epso.am/poweren.htm'}
fetch_exchange(AM, GE) ->
{'sortedZoneKeys': 'AM->GE', 'datetime': datetime.datetime(2019, 4, 28, 14, 35, 27, 415214, tzinfo=tzfile('Asia/Yerevan')), 'netFlow': 5.9, 'source': 'http://epso.am/poweren.htm'}
fetch_exchange(AM, IR) ->
{'sortedZoneKeys': 'AM->IR', 'datetime': datetime.datetime(2019, 4, 28, 14, 35, 30, 852453, tzinfo=tzfile('Asia/Yerevan')), 'netFlow': 95.8, 'source': 'http://epso.am/poweren.htm'}
```

**Comment by systemcatch** [comment]
For connection drops the EM server might already handle resending requests, @corradio?

**Comment by systemcatch** [comment]
Working ok on the mockserver.

AM

**Comment by alixunderplatz** [comment]
@corradio @systemcatch is the AM<->GE exchange fetched twice in the respective parsers for AM and GE? could this be problematic or is this irrelevant due to different timestamps?

**Comment by corradio** [comment]
@systemcatch the back runs the parsers every 5min. If it fails then it will be just retry 5min after.
@alixunderplatz the exchange is only fetched once, using the parser specified in exchanges.json

**Comment by systemcatch** [comment]
@alixunderplatz at the moment only the fetch_exchange function in the GE parser will run since that is the only one setup in the exchange config.

**Comment by systemcatch** [comment]
We're good to go I think.

**Comment by corradio** [comment]
Great - let's do it!

**Comment by brunolajoie** [comment]
Congrats and many thanks everyone! 🙏
